batch/batch/driver/instance_collection/pool.py

```diff
@@ -418,8 +433,18 @@ async def create_instances(self):
                     break
 
         n_live_instances = self.n_instances_by_state['pending'] + self.n_instances_by_state['active']
-        if self.enable_standing_worker and n_live_instances == 0 and self.max_instances > 0:
-            await self.create_instance(
+        n_standing_instances_needed = max(0, self.min_instances - self.n_instances)
+        n_standing_instances_needed = min(
+            n_standing_instances_needed,
+            self.max_live_instances - n_live_instances,
+            self.max_instances - self.n_instances,
+            remaining_instances_per_autoscaler_loop,
+            # 20 queries/s; our GCE long-run quota
+            300,
+        )
```

Comment:
I looked at this pretty carefully and think it's correct but since you asked me to double check I thought a bit to see if maybe we can write this in a clearer way. I am also always a bit foggy on live vs total instances so let me know if I'm misunderstanding something. Also, this is just an alternative. Feel free to take it if you think it's clearer or just dismiss and I'll approve.

```python3
capacity_for_live_instances = max(0, self.max_live_instances - n_live_instances)
capacity_for_any_instances = max(0, self.max_instances, self.n_instances)
capacity_for_new_instances = min(capacity_for_live_instances, capacity_for_any_instances)
if capacity_for_new_instances > 0:
    n_standing_instances_to_provision = min(capacity_for_new_instances, self.min_instances, remaining_instances_per_autoscaler_loop, 300)
    if n_standing_instances_to_provision > 0:
        await self._create_instances(…)
```

Reply:
This is much clearer! I had to modify it slightly though because you need `max(0, self.min_instances - self.n_instances)` in there to know how many you need to create to get to the minimum.